services/app-api/src/resolvers/rate/withdrawRate.test.ts

```diff
@@ -0,0 +1,275 @@
+import {
+    iterableNonCMSUsersMockData,
+    testCMSUser,
+    testStateUser,
+} from '../../testHelpers/userHelpers'
+import { constructTestPostgresServer } from '../../testHelpers/gqlHelpers'
+import {
+    createAndSubmitTestContractWithRate,
+    createAndUpdateTestContractWithRate,
+    submitTestContract,
+    unlockTestContract,
+} from '../../testHelpers/gqlContractHelpers'
+import { WithdrawRateDocument } from '../../gen/gqlClient'
+import { mockStoreThatErrors } from '../../testHelpers/storeHelpers'
+
+describe('withdrawRate', () => {
+    it('can withdraw a rate without errors', async () => {
+        const stateUser = testStateUser()
+        const cmsUser = testCMSUser()
+
+        const stateServer = await constructTestPostgresServer({
+            context: {
+                user: stateUser,
+            },
+        })
+
+        const cmsServer = await constructTestPostgresServer({
+            context: {
+                user: cmsUser,
+            },
+        })
+
+        const contract = await createAndSubmitTestContractWithRate(stateServer)
+        const rateID = contract.packageSubmissions[0].rateRevisions[0].rateID
+
+        const withdrawnRateResult = await cmsServer.executeOperation({
+            query: WithdrawRateDocument,
+            variables: {
+                input: {
+                    rateID: rateID,
+                    updatedReason: 'Withdraw invalid rate',
+                },
+            },
+        })
+
+        const withdrawnRate = withdrawnRateResult.data?.withdrawRate.rate
+
+        // expect no errors
+        expect(withdrawnRateResult.errors).toBeUndefined()
+
+        // expect withdrawn status
+        expect(withdrawnRate).toEqual(
+            expect.objectContaining({
+                reviewStatus: 'WITHDRAWN',
+                consolidatedStatus: 'WITHDRAWN',
+            })
+        )
+
+        // expect correct actions
+        expect(withdrawnRate.reviewStatusActions).toHaveLength(1)
+        expect(withdrawnRate.reviewStatusActions).toEqual(
+            expect.arrayContaining([
+                expect.objectContaining({
+                    updatedAt: expect.any(Date),
+                    updatedBy: expect.objectContaining({
+                        role: cmsUser.role,
+                        email: cmsUser.email,
+                        givenName: cmsUser.givenName,
+                        familyName: cmsUser.familyName,
+                    }),
+                    updatedReason: 'Withdraw invalid rate',
+                    actionType: 'WITHDRAW',
+                    rateID,
+                }),
+            ])
+        )
+    })
+    it('returns error if rate is in invalid status to withdraw', async () => {
```

Comment:
I know we do these workflows a lot in our tests, but maybe we can refactor to something like this to make it a little easier to see where each thing is happening:

```typescript
    describe('withdrawRate invalid status handling', () => {
        let stateUserInstance: ApolloServer
        let cmsUserInstance: ApolloServer
        let rateID: string
        let contractID: string

        beforeEach(async () => {
            stateUserInstance = await constructTestPostgresServer({
                context: { user: testStateUser() },
            })
            cmsUserInstance = await constructTestPostgresServer({
                context: { user: testCMSUser() },
            })
            // Set up base contract/rate
            const draftContract =
                await createAndUpdateTestContractWithRate(stateUserInstance)
            rateID = draftContract.draftRates?.[0].id as string
            contractID = draftContract.id
        })

        it('cannot withdraw a draft rate', async () => {
            const result = await cmsUserInstance.executeOperation({
                query: WithdrawRateDocument,
                variables: {
                    input: {
                        rateID,
                        updatedReason: 'Withdraw draft rate',
                    },
                },
            })
            expect(result.errors?.[0].message).toBe(
                'Attempted to withdraw rate with wrong status: DRAFT'
            )
        })

        it('cannot withdraw an unlocked rate', async () => {
            await submitTestContract(stateUserInstance, contractID)
            await unlockTestContract(
                cmsUserInstance,
                contractID,
                'Unlock contract'
            )

            const failedWithdrawUnlockedRate =
                await cmsUserInstance.executeOperation({
                    query: WithdrawRateDocument,
                    variables: {
                        input: {
                            rateID,
                            updatedReason: 'Withdraw unlocked rate',
                        },
                    },
                })

            // expect error for attempting to withdraw an unlocked rate
            expect(failedWithdrawUnlockedRate.errors?.[0]).toBeDefined()
            expect(failedWithdrawUnlockedRate.errors?.[0].message).toBe(
                'Attempted to withdraw rate with wrong status: UNLOCKED'
            )
        })

        it('cannot withdraw an already withdrawn rate', async () => {
            await submitTestContract(stateUserInstance, contractID)
            await unlockTestContract(
                cmsUserInstance,
                contractID,
                'Unlock contract'
            )
            await submitTestContract(stateUserInstance, contractID, 'Resubmit')
            await cmsUserInstance.executeOperation({
                query: WithdrawRateDocument,
                variables: {
                    input: {
                        rateID,
                        updatedReason: 'Withdraw submitted rate',
                    },
                },
            })

            const failedWithdrawWithdrawnRate =
                await cmsUserInstance.executeOperation({
                    query: WithdrawRateDocument,
                    variables: {
                        input: {
                            rateID,
                            updatedReason: 'Withdraw already withdrawn rate',
                        },
                    },
                })

            // expect error for attempting to withdraw a withdrawn rate
            expect(failedWithdrawWithdrawnRate.errors?.[0]).toBeDefined()
            expect(failedWithdrawWithdrawnRate.errors?.[0].message).toBe(
                'Attempted to withdraw rate with wrong status: WITHDRAWN'
            )
        })
    })
    ```

Reply:
Interesting setup for the tests. We usually setup the servers and data inside the tests themselves for isolation since the data is saved to the test DB and persistent between tests. I think the `beforeEach` would also work since we run test sequentially, at least locally. If we were to run any of these tests in parallel it would fail.

I'm thinking we move these tests into the separate block like you have, but keep the setup code in the tests for isolation. Let me know what you think, there could be some kind of middle ground that would still preserve test isolation.